MatrixSDK/Data/MXRoom.m

```diff
@@ -335,21 +451,600 @@ - (MXHTTPOperation*)sendStateEventOfType:(MXEventTypeString)eventTypeString
+                    
+                    MXHTTPOperation *operation2 = [self sendMessageWithContent:msgContent localEcho:&event success:success failure:failure];
+                    if (operation2)
+                    {
+                        // Mutate MXHTTPOperation so that the user can cancel this new operation
+                        [operation mutateTo:operation2];
+                    }
+                    
+                } failure:onFailure];
+            } failure:onFailure];
+        }
+        else
+        {
+            // Upload thumbnail
+            [thumbUploader uploadData:videoThumbnailData filename:nil mimeType:@"image/jpeg" success:^(NSString *thumbnailUrl) {
+                
+                // Upload video
+                NSData* videoData = [NSData dataWithContentsOfFile:convertedLocalURL.path];
+                if (videoData)
+                {
+                    // Copy the cached thumbnail to the actual cacheFile path
+                    NSString *absoluteURL = [self.mxSession.matrixRestClient urlOfContent:thumbnailUrl];
+                    NSString *actualCacheFilePath = [MXMediaManager cachePathForMediaWithURL:absoluteURL andType:@"image/jpeg" inFolder:self.roomId];
+                    NSError *error;
+                    [[NSFileManager defaultManager] copyItemAtPath:cacheFilePath toPath:actualCacheFilePath error:&error];
+                    
+                    MXMediaLoader *videoUploader = [MXMediaManager prepareUploaderWithMatrixSession:self.mxSession initialRange:0.1 andRange:0.9];
+                    
+                    // Create a fake file name based on videoData to keep the same name for the same file.
+                    NSString *dataHash = [videoData mx_MD5];
+                    if (dataHash.length > 7)
+                    {
+                        // Crop
+                        dataHash = [dataHash substringToIndex:7];
+                    }
+                    NSString *extension = [MXTools fileExtensionFromContentType:mimetype];
+                    NSString *filename = [NSString stringWithFormat:@"video_%@%@", dataHash, extension];
+                    msgContent[@"body"] = filename;
+                    
+                    // Update thumbnail URL with the actual mxc: URL
+                    msgContent[@"info"][@"thumbnail_url"] = thumbnailUrl;
+                    
+                    // Apply the nasty trick again so that the cell can monitor the upload progress
+                    msgContent[@"url"] = videoUploader.uploadId;
+                    
+                    [self updateOutgoingMessage:event.eventId withOutgoingMessage:event];
+                    
+                    // Force a refresh of the displayed echo by posting sent state change notification even if the state did not change (it is still uploading)
+                    [[NSNotificationCenter defaultCenter] postNotificationName:kMXEventDidChangeSentStateNotification object:event userInfo:nil];
```

Comment:
Simple remark: it may worth to have a processing, converting or compressing state in MXEventSentState because sending this "false" notif  looks like hacky here.